include/mixer.h

```diff
@@ -93,6 +98,8 @@ enum LINE_INDEX : uint8_t {
 	// standard at the host-level, then additional line indexes would go here.
 };
 
+enum class FilterState { Off, On, ForcedOn };
```

Comment:
Might as well carrying the values in a single-byte type:

``` c++
enum class FilterState : uint8_t { Off, On, ForcedOn };
```

The other benefit of using an unsigned type is it lets us use them in cross-platform bitfields (if we ever want to go there), here's a toy example:

``` c++
struct filter_states {
   FilterState disney : 2;
   FilterState innovation : 2;
   FilterState gus : 2;
   FilterState opl : 2;
   FilterState pcspeaker : 2;
   FilterState ps1 : 2;
   FilterState sb : 2;
   FilterState tandy : 2;
};

// use it like any other struct:

filter_states.sb = FilterState::ForcedOn;
filter_states.ps1 = FilterState::Off;
...

// entire struct is only 2 bytes!
assert(sizeof(filter_states) == 2);
```

Reply:
I didn't know that bitfield trick, that's interesting.

To be honest, I'm not really a fan of overriding the default width unless we really need it, as it just adds extra noise; it's just not worth it in my view. Plus, to me at least, it signifies that something special is going on and this is not just a regular enum.

I'd personally use and expect to see the width explicitly specified in these cases:

- you need it for C interop and/or some low-level bit-twiddling
- you have many many thousands or millions of these in a container of packed structs
- the enum is part of some binary data format that gets written to disk or sent through the network (subclass of the packed struct case)

Your hypothetical example can fall into either of these categories. But because we're not writing the enum to disk, we can always add the explicit width specifier later -- if we ever need to, that is. Otherwise, adding it feels to me as unnecessary detail and noise that doesn't really have much benefit. I wouldn't even call it micro-optimisation because it doesn't optimise anything 😎 

Most likely there wouldn't be any memory savings because of memory alignment requirements. Everything narrower than 64-bits would effectively use up to 4 bytes in memory on x64 with the extra padding bytes. It depends on the order of fields in the structs because struct member reordering is _streng verboten_ in all C/C++ compilers in existence (for C interop and because of unions). I wrote about it [here](https://blog.johnnovak.net/2020/12/21/nim-apocrypha-vol.-i/#16-nbsp-saving-memory-by-object-field-reordering), and indeed, you can reduce memory requirements by several folds by packing structs in large matrixes and such. But for just a handful of enums, who cares really? We're not coding for embedded systems with 4k of RAM. We're using ~6 megs worth of DLL fiels already, so what are a few extra bytes 😛 

Anyway, it's a minor point, just wanted to share my view. I tend to go with defaults everywhere, unless it makes a tangible difference -- otherwise it's just more letters on the screen and potentially confusing.

Reply:
> To be honest, I'm not really a fan of overriding the default width unless we really need it, as it just adds extra noise; it's just not worth it in my view. Plus, to me at least, it signifies that something special is going on and this is not just a regular enum.

I fully agree (especially when things get noisy). 

I have a background in science and performance-sensitive applications, back when 4 GB of RAM was a lot for a server -- small changes like cutting disk IO in half, or fitting twice as much data in memory (say, working with floats instead of doubles) could more than double performance (as a trivial example). So this is some of that old mindset leaking through (with zero-benefit here :laughing:) 

I do like to constraints some type sizes when we're certainly the values they hold should never grow beyond some point - it shares some information with readers that we never expect massive values.  Likewise, if I'm using a 64bit counter, it conveys information about that too.  But those are both extreme ends.. if we can use a basic `int` for most cases, that's perfect!

Reply:
> I have a background in science and performance-sensitive applications, back when 4 GB of RAM was a lot for a server

Noice! My background is learning x86 assembly & VGA programming as a kid because of games and the demoscene (on a 486 DX2-66, and probably I had 4 or 8 megs or RAM, but I was constrained by 64k segments in real mode before I learned about 32-bit flat mode). I actually learned asm several years before C 😎 Then I worked for a few years maintaining some voxel-based rendering engine written in C. I always wanted to be a graphics programmer, but somehow ended up doing server-side stuff, so it's been Java, Scala & Kotlin for me in the past ~15 years professionally, then I use Nim in my hobby projects.

Many times I still tend to have a performance-first bias at work, but my colleagues have been quite successful at conditioning my micro-optimisation tendencies out of me 😂 Most of the time the DB or the network is the bottleneck in  my work anyway.

Choosing the most appropriate type is a good idea, and I certainly approve of that (although mixing signed & unsigned integers can get tricky, but I think the compilers give appropriate warnings about such cases these days).

Reply:
Sorry for the late reply here! Very cool and fun background - I was merely enjoying the fruits of mod and demo scene in the early 90s, that's awesome you were tinkering with your own assembly and demo scene code back then!

We're all definitely on the same page there :-) 

>  I always wanted to be a graphics programmer,

@Wengier's done the initial port of voodoo on a branch. When we eventually review and merge that, then we can start looking at improving and refactoring it too - and I'm positive you'll find lots of fun 3D stuff in there (3D API calls emulated in code.. this is where every optimization has payoffs!)